### PR TITLE
Curl exploit

### PR DESCRIPTION
**Pull request by shaun-mac**
This totally won't cause our snakes to just … give up …

## Testing notes

- You'll need to have `redis` installed on your machine to run this locally. Open a terminal and run `redis-server` to start the Redis instance.

### CODE REVIEW

**Review by DevYves — COMMENTED** [review]
I love this idea! However, I think it might be a bit tricky because we are resetting the giveUp variable at the start of every game... so the only way for someone to use this exploit would be to curl us in the middle of a game. Which they could do but is sort of tight timing? However, we can easily add a persistent layer on top of this... assuming.. time and stuff and if not we have a second exploit that is possible for someone to use.💪

**Comment by mike-anderson** [comment]
> I love this idea! However, I think it might be a bit tricky because we are resetting the giveUp variable at the start of every game... so the only way for someone to use this exploit would be to curl us in the middle of a game. Which they could do but is sort of tight timing? However, we can easily add a persistent layer on top of this... assuming.. time and stuff and if not we have a second exploit that is possible for someone to use.💪

I agree, the games may even be run before the arena is returned depending on the load on the servers apparently. Safest would be that the give up variable is pushed ahead of time. We would have to restart or reset after each match

or the variable would need to be set as a lookup against a snake name

**Comment by shaun-mac** [comment]
> > I love this idea! However, I think it might be a bit tricky because we are resetting the giveUp variable at the start of every game... so the only way for someone to use this exploit would be to curl us in the middle of a game. Which they could do but is sort of tight timing? However, we can easily add a persistent layer on top of this... assuming.. time and stuff and if not we have a second exploit that is possible for someone to use.💪
> 
> I agree, the games may even be run before the arena is returned depending on the load on the servers apparently. Safest would be that the give up variable is pushed ahead of time. We would have to restart or reset after each match
> 
> or the variable would need to be set as a lookup against a snake name

Fair points, @DevYves @mike-anderson 

@samueljohnwilson and I chatted about this – he had mentioned that the game may 'catch' up to the point where the opponent is able to send the curl mid-game. Another option we discussed is having the endpoint display a button that they could press (which might make the action faster/easier to do mid game). This is honestly sort of a first pass, and I have started looking into persistent data storage as a follow-up.

File: src/index.ts [hunk 1]
```diff
@@ -71,6 +97,18 @@ app.get('/version', (_, response) => {
   return response.send(process.env.VERSION || 'undefined');
 });
 
+app.post('/shout', urlencodedParser, (request, response) => {
```
Comment (samueljohnwilson) [inline src/index.ts]: Given that we're not planning to use the `shout` in the API response to trigger the vulnerability anymore, I wonder if we might want to modify the naming convention here so as not to confuse the two?

Comment (mike-anderson) [reply src/index.ts]: if it's not too confusing I think it might be good to have both paths work to trigger the same behavior

Comment (shaun-mac) [reply src/index.ts]: @samueljohnwilson Renamed in e3b6fc1

**Comment by shaun-mac** [comment]
@brandonb927 Adding you to the PR – please let me know what updates need to be made (if any) when the redis-server droplet is deployed.

**Review by samueljohnwilson — APPROVED** [review]

File: src/index.ts [hunk 2]
```diff
@@ -6,7 +6,20 @@ import {
   genericErrorHandler,
   poweredByHandler,
 } from './handlers';
-import SnakeBrain from './SnakeBrain';
+import { getNemesis } from './helpers';
+import SnakeBrain from '../src/SnakeBrain';
+import { IGameState } from '../src/Types';
+import redis from 'redis';
+
+const client = redis.createClient();
```
Comment (brandonb927) [inline src/index.ts]:
```suggestion
const REDIS_HOST = process.env.REDIS_HOST || 'localhost';
const REDIS_PORT = process.env.REDIS_PORT || 6379;
const client = redis.createClient(REDIS_PORT, REDIS_HOST);
```

**Pull request merged**